src/memory.rs

```diff
@@ -54,100 +54,138 @@ impl ByteSliceView {
 }
 
-impl Buffer {
-    /// `read` provides a reference to the included data to be parsed or copied elsewhere
-    ///
-    /// # Safety
-    ///
-    /// The caller must make sure that the `Buffer` points to valid and initialized memory
-    pub unsafe fn read(&self) -> Option<&[u8]> {
-        if self.ptr.is_null() {
-            None
-        } else {
-            Some(slice::from_raw_parts(self.ptr, self.len))
+impl UnmanagedVector {
+    /// Consumes this optional vector for manual management.
+    /// This is a zero-copy operation.
+    pub fn new(source: Option<Vec<u8>>) -> Self {
+        match source {
+            Some(data) => {
+                let mut data = mem::ManuallyDrop::new(data);
+                Self {
+                    is_nil: false,
+                    ptr: data.as_mut_ptr(),
+                    len: data.len(),
+                    cap: data.capacity(),
+                }
+            }
+            None => Self {
+                is_nil: true,
+                ptr: std::ptr::null_mut::<u8>(),
+                len: 0,
+                cap: 0,
+            },
         }
     }
 
-    /// consume must only be used on memory previously released by from_vec
-    /// when the Vec is out of scope, it will deallocate the memory previously referenced by Buffer
-    ///
-    /// # Safety
-    ///
-    /// if not empty, `ptr` must be a valid memory reference, which was previously
-    /// created by `from_vec`. You may not consume a slice twice.
-    /// Otherwise you risk double free panics
-    pub unsafe fn consume(self) -> Vec<u8> {
-        if self.ptr.is_null() {
-            vec![]
+    pub fn len(&self) -> Option<usize> {
+        if self.is_nil {
+            None
         } else {
-            Vec::from_raw_parts(self.ptr, self.len, self.cap)
+            Some(self.len)
         }
     }
 
-    /// Creates a new zero length Buffer with the given capacity
-    pub fn with_capacity(capacity: usize) -> Self {
-        Buffer::from_vec(Vec::<u8>::with_capacity(capacity))
+    pub fn is_none(&self) -> bool {
+        self.is_nil
     }
 
-    // this releases our memory to the caller
-    pub fn from_vec(v: Vec<u8>) -> Self {
-        v.into()
+    pub fn is_some(&self) -> bool {
+        !self.is_none()
+    }
+
+    /// Takes this UnmanagedVector and turns it into a regular, managed Rust vector.
+    /// Calling this on two copies of UnmanagedVector leads to double free crashes.
+    pub fn consume(self) -> Option<Vec<u8>> {
+        if self.is_nil {
+            None
+        } else {
+            Some(unsafe { Vec::from_raw_parts(self.ptr, self.len, self.cap) })
+        }
+    }
+
+    /// Drops the underlying Vector and converts this instance to None
+    pub fn drop_vector(&mut self) {
+        if !self.is_nil {
+            let _ = unsafe { Vec::from_raw_parts(self.ptr, self.len, self.cap) };
+            self.is_nil = true;
+        }
     }
 }
 
-impl Default for Buffer {
+impl Default for UnmanagedVector {
```

Comment:
I like this 0 alloc default

Reply:
By the way, do you know that Rust does not allocate heap as long as the vector capacity is 0? So `Vec::new()` does not allocate. `Vec::with_capacity(0)` does not allocate. But `Vec::with_capacity(1)` does.